src/connector/src/sink/mongodb.rs

```diff
@@ -665,66 +635,61 @@ impl MongodbPayloadWriter {
                 .add_delete(pk)?,
         }
-
-        self.buffered_entries += 1;
-        if self.buffered_entries >= self.max_entries {
-            self.flush_upsert(upsert_builder).await?;
-        }
         Ok(())
     }
 
-    async fn flush_insert(
-        &mut self,
+    fn flush_insert(
+        &self,
         insert_builder: &mut HashMap<MongodbNamespace, InsertCommandBuilder>,
-    ) -> Result<()> {
+    ) -> Result<Vec<impl futures::Future<Output = std::result::Result<(), SinkError>>>> {
         // TODO try sending bulk-write of each collection concurrently to improve the performance when
         // `dynamic collection` is enabled. We may need to provide best practice to guide user on setting
         // the MongoDB driver's connection properties.
+        let mut futures = Vec::with_capacity(insert_builder.len());
         for (ns, builder) in insert_builder.drain() {
-            self.send_bulk_write_command(&ns.0, builder.build()).await?;
+            let db = self.client.database(&ns.0);
+            futures.push(Self::send_bulk_write_command(db, builder.build()));
         }
-        self.buffered_entries = 0;
-        Ok(())
+        Ok(futures)
     }
 
-    async fn flush_upsert(
-        &mut self,
+    fn flush_upsert(
+        &self,
         upsert_builder: &mut HashMap<MongodbNamespace, UpsertCommandBuilder>,
-    ) -> Result<()> {
+    ) -> Result<Vec<impl futures::Future<Output = std::result::Result<(), SinkError>>>> {
         // TODO try sending bulk-write of each collection concurrently to improve the performance when
         // `dynamic collection` is enabled. We may need to provide best practice to guide user on setting
         // the MongoDB driver's connection properties.
+        let mut futures = Vec::with_capacity(upsert_builder.len());
         for (ns, builder) in upsert_builder.drain() {
             let (upsert, delete) = builder.build();
             // we are sending the bulk upsert first because, under same pk, the `Insert` and `UpdateInsert`
             // should always appear before `Delete`. we have already ignored the `UpdateDelete`
             // which is useless in upsert mode.
+            let db = self.client.database(&ns.0);
             if let Some(upsert) = upsert {
-                self.send_bulk_write_command(&ns.0, upsert).await?;
+                futures.push(Self::send_bulk_write_command(db.clone(), upsert));
             }
             if let Some(delete) = delete {
-                self.send_bulk_write_command(&ns.0, delete).await?;
+                futures.push(Self::send_bulk_write_command(db, delete));
             }
         }
-        self.buffered_entries = 0;
-        Ok(())
+        Ok(futures)
     }
 
-    async fn send_bulk_write_command(&self, database: &str, command: Document) -> Result<()> {
-        let db = self.client.database(database);
-
+    async fn send_bulk_write_command(db: Database, command: Document) -> Result<()> {
```

Comment:
I just realize that both `flush_upsert` and `flush_insert` are returning a vector of the same future generated by this method.

If so, we can make the return type of `flush_upsert` and `flush_insert` to be the same, and then we don't need to add a boxed to the try_join_all future.

This method can be defined in this way
```
type SendBulkWriteCommandFuture = impl Future<Output = Result<()>> + 'static;

fn send_bulk_write_command(db: Database, command: Document) -> SendBulkWriteCommandFuture {
   async move {
        ...
    }
}
```

And then both `flush_upsert` and `flush_insert` can change to return `TryJoinAll<SendBulkWriteCommandFuture>`, and in `write_chunk` we don't need an extra `boxed` to unify the type.